src/platforms/esp32/main/spi_driver.c

```diff
@@ -362,6 +370,153 @@ static term spidriver_write_at(Context *ctx, term req)
+        }
+        size_t write_bits = (size_t) term_to_int(write_bits_term);
+        if (write_bits > binary_bits) {
+            ESP_LOGE(TAG, "More write bits specified (%u) than are available (%u)", write_bits, binary_bits);
+            return BADARG_ATOM;
+        }
+        transaction->length = write_bits;
+    }
+
+    if (!write_only) {
+        term read_bits_term = term_get_map_assoc_default(ctx, transaction_term, context_make_atom(ctx, read_bits_atom), binary_bits_term);
+        if (!term_is_integer(read_bits_term)) {
+            ESP_LOGE(TAG, "read_bits transaction entry is not an integer");
+            return BADARG_ATOM;
+        }
+        size_t read_bits = (size_t) term_to_int(read_bits_term);
+        transaction->rxlength = read_bits;
+        *output_size = (read_bits % 8) == 0 ? (read_bits / 8) : ((read_bits / 8) + 1);
+    }
+
+    return OK_ATOM;
+}
+
+//
+// write operations
+//
+
+static term spidriver_write(Context *ctx, term req)
+{
+    TRACE("spidriver_write\n");
+    struct SPIData *spi_data = ctx->platform_data;
+
+    // cmd is at index 0
+    term device_term = term_get_tuple_element(req, 1);
+    term transaction_term = term_get_tuple_element(req, 2);
+
+    struct SPIDevice *device = get_spi_device(spi_data, device_term);
+    if (IS_NULL_PTR(device)) {
+        if (UNLIKELY(memory_ensure_free(ctx, 3) != MEMORY_GC_OK)) {
+            return OUT_OF_MEMORY_ATOM;
+        }
+        return create_pair(ctx, ERROR_ATOM, context_make_atom(ctx, device_not_found_atom));
+    }
+
+    struct spi_transaction_t transaction = { 0 };
+    size_t output_size = 0;
+    term err_term = populate_transaction(ctx, &transaction, transaction_term, true, &output_size);
+    if (err_term != OK_ATOM) {
+        ESP_LOGE(TAG, "Invalid transaction");
+        if (UNLIKELY(memory_ensure_free(ctx, 3) != MEMORY_GC_OK)) {
+            return OUT_OF_MEMORY_ATOM;
+        }
+        return create_pair(ctx, ERROR_ATOM, err_term);
+    }
+
+    // TODO replace spi_device_polling_transmit with a interrupt-based mechanism
+    esp_err_t err = spi_device_polling_transmit(device->handle, &transaction);
+    if (UNLIKELY(err != ESP_OK)) {
+        ESP_LOGE(TAG, "spidriver_write failed with err=%i", err);
+        if (UNLIKELY(memory_ensure_free(ctx, 3) != MEMORY_GC_OK)) {
+            return OUT_OF_MEMORY_ATOM;
+        }
+        return create_pair(ctx, ERROR_ATOM, term_from_int(err));
+    }
+    return OK_ATOM;
+}
+
+//
+// write_read operations
+//
+
+static term spidriver_write_read(Context *ctx, term req)
+{
+    TRACE("spidriver_write_read\n");
+    struct SPIData *spi_data = ctx->platform_data;
+
+    // cmd is at index 0
+    term device_term = term_get_tuple_element(req, 1);
+    term transaction_term = term_get_tuple_element(req, 2);
+
+    struct SPIDevice *device = get_spi_device(spi_data, device_term);
+    if (IS_NULL_PTR(device)) {
+        if (UNLIKELY(memory_ensure_free(ctx, 3) != MEMORY_GC_OK)) {
```

Comment:
let's create a device_not_found_error() function, so we can avoid copying&pasting this, and we can just do:
```c
if (IS_NULL_PTR(device)) {
    return device_not_found_error(ctx);
}
```